independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/ListCategoriesCommandHandler.java

```diff
@@ -0,0 +1,96 @@
+package io.quarkus.devtools.commands.handlers;
+
+import io.quarkus.devtools.commands.ListCategories;
+import io.quarkus.devtools.commands.data.QuarkusCommandException;
+import io.quarkus.devtools.commands.data.QuarkusCommandInvocation;
+import io.quarkus.devtools.commands.data.QuarkusCommandOutcome;
+import io.quarkus.devtools.messagewriter.MessageWriter;
+import io.quarkus.devtools.project.BuildTool;
+import io.quarkus.registry.catalog.Category;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.function.BiConsumer;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Instances of this class are thread-safe. It lists extension categories according to the options passed in as properties of
+ * {@link QuarkusCommandInvocation}
+ */
+public class ListCategoriesCommandHandler implements QuarkusCommandHandler {
+
+    private static final String ID_FORMAT = "%-20s";
+    private static final String FULL_FORMAT = "%-30s %-20s %s";
+
+    private static final String LIST_EXTENSIONS_HINT = "\nTo list extensions in given category, use:\n%s";
+    private static final String MORE_INFO_HINT = "\nTo get more information, append `%s` to your command line.";
+
+    @Override
+    public QuarkusCommandOutcome execute(QuarkusCommandInvocation invocation) throws QuarkusCommandException {
+
+        final MessageWriter log = invocation.log();
+        final boolean cli = invocation.getValue(ListCategories.FROM_CLI, false);
+        final boolean batchMode = invocation.getValue(ListCategories.BATCH_MODE, false);
+        final String format = invocation.getValue(ListCategories.FORMAT, "");
+
+        final Collection<Category> categories = invocation.getExtensionsCatalog().getCategories();
+
+        if (!cli) {
```

Comment:
Why is this a cli mode check instead of a batch mode check?

Reply:
Well, I'm not absolutely sure (I tried to do the same thing that `ListExtensionsCommandHandler` does), but I think the original intention was this to give a user some introductory text like "Here starts the output you wanted:" in case that the handler is run from Gradle or Maven command (`cli == false`), as the output would be drowned in the usual verbose Maven/Gradle log messages.

On the other hand when the handler is run from the CLI, it would print the list directly, as there is no noise around it (apart from the possible "hints" at the end).

I think the behavior is sensible, but I don't have anything against hooking this to the `batchMode` variable instead.